s3dol/base.py

```diff
@@ -0,0 +1,227 @@
+    try:
+        return get_client('environment variables', **session_kwargs)
+    except S3DolException:
+        pass
+    session = boto3.Session(**session_kwargs)
+    if session.get_credentials() is None:
+        raise noCredentialsFound
+    client = session.client('s3', endpoint_url=endpoint_url)
+    return client
+
+
+@dataclass
+class BaseS3BucketReader(dol.base.KvReader):
+    client: boto3.client
+    bucket_name: str
+
+    def __iter__(self):
+        _obj_list = self.client.list_objects(Bucket=self.bucket_name)
+        return (o['Key'] for o in _obj_list.get('Contents', []))
+
+    def __getitem__(self, k: str):
+        return self.client.get_object(Bucket=self.bucket_name, Key=k)['Body'].read()
+
+
+class BaseS3BucketDol(BaseS3BucketReader, dol.base.KvPersister):
+    def __setitem__(self, k, v):
+        self.client.put_object(Bucket=self.bucket_name, Key=k, Body=v)
+
+    def __delitem__(self, k):
+        self.client.delete_object(Bucket=self.bucket_name, Key=k)
+
+
+@dataclass
+class S3BucketReader(BaseS3BucketReader):
+    prefix: str = ''
+    delimiter: str = '/'
+
+    def _id_of_key(self, k):
+        return f'{self.prefix}{k}'
+
+    def _key_of_id(self, id):
+        return id[len(self.prefix) :]
+
+    def __iter__(self):
+        return (
+            self._key_of_id(_id)
+            for _id in super().__iter__()
+            if _id.startswith(self.prefix)
+        )
+
+    def __getitem__(self, k: str):
+        _id = self._id_of_key(k)
+        if _id.endswith(self.delimiter):
+            _kw = {**self.__dict__, 'prefix': _id}
+            return type(self)(**_kw)
+        return super().__getitem__(_id)
+
+
+class S3BucketDol(S3BucketReader, BaseS3BucketDol):
+    def __setitem__(self, k, v):
+        _id = self._id_of_key(k)
+        return super().__setitem__(_id, v)
+
+    def __delitem__(self, k):
+        _id = self._id_of_key(k)
+        return super().__delitem__(_id)
+
+
+class S3ClientReader(dol.base.KvReader):
+    def __init__(
+        self, *, s3_bucket_dol=S3BucketDol, profile_name=None, **session_kwargs
+    ):
+        self.client = get_client(profile_name=profile_name, **session_kwargs)
+        self.s3_bucket_dol = s3_bucket_dol
+
+    def __iter__(self):
+        return (b['Name'] for b in self.client.list_buckets().get('Buckets', []))
+
+    def __getitem__(self, k: str):
+        return self.s3_bucket_dol(client=self.client, bucket_name=k)
+
+
+class S3ClientDol(S3ClientReader, dol.base.KvPersister):
+    def __init__(
+        self, *, s3_bucket_dol=S3BucketDol, profile_name=None, **session_kwargs
+    ):
+        super().__init__(
+            s3_bucket_dol=s3_bucket_dol, profile_name=profile_name, **session_kwargs
+        )
+
+    def __getitem__(self, k: str):
+        """Get bucket. If bucket does not exist, create it.
+
+        :param k: bucket name
+        :type k: str
+        :return: S3 bucket dol
+        """
+        self.client.create_bucket(Bucket=k)
```

Comment:
Not sure we want to create it in that case. `__get_item__` should return `None` if k does not exist

Reply:
@valentin-feron  As stated above, I think the final S3Dol class should have everything included and the other classes can be used if want less. It is simple to make a version without bucket creation and throws and error like:  
```
S3DolWriteInBucketOnly = functools.partial(S3Dol, s3_client_dol=S3ClientReader)
```
You could make the case that create bucket is too dangerous and should explicitly use the class with more writing capabilities.  That usage could also be restricted by the access keys instead of the python code.

Reply:
keeping as is